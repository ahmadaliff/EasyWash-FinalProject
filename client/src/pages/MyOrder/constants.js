export const GET_ORDERS = 'myOrder/GET_ORDERS';
export const SET_ORDERS = 'myOrder/SET_ORDERS';
export const RESET_ORDERS = 'myOrder/RESET_ORDERS';
export const CANCEL_ORDER = 'myOrder/CANCEL_ORDER';
