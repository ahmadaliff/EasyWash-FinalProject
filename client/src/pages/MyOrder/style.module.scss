@import '@styles/base';

.mainWrap {
  @include directionY;
  padding: 1rem;
  @include apply-at-md-size {
    padding: 2rem 5rem 2rem 5rem;
  }
  min-height: calc(100vh - 3.55rem);
  align-items: center;
  box-sizing: border-box;
  position: relative;
  * {
    box-sizing: border-box;
  }
  .header {
    @include directionX;
    justify-content: space-between;
    width: 100%;
    align-items: center;
    .backButton {
      margin-left: auto;
      background-color: var(--color-primary);
      border-radius: 10rem;
      padding: 0.25rem 1rem 0.25rem 1rem;
      text-transform: capitalize;
      font-size: medium;
      &:hover {
        background-color: var(--color-primary);
      }
    }
  }
  .order {
    @include directionX;
    align-items: center;
    gap: 0.125rem;
    padding: 0.25rem;
    width: 100%;
    .orderInfoWrap {
      width: 100%;
      p {
        margin: 0;
        font-size: small;
      }
      .quantity {
        @include directionX;
        align-items: center;
        justify-content: space-between;
        font-size: small;
      }
      .orderInfo {
        @include directionX;
        justify-content: space-between;
      }
      .buttonDetail {
        font-size: small;
        color: var(--color-primary);
        font-weight: 600;
        &:hover {
          cursor: pointer;
          transform: scale(1.0125, 1.0125);
          text-decoration: underline;
        }
      }
    }
  }
}
