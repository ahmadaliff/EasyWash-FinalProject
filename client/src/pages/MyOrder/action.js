import { CANCEL_ORDER, GET_ORDERS, RESET_ORDERS, SET_ORDERS } from './constants';

export const actionGetMyOrder = () => ({
  type: GET_ORDERS,
});

export const actionSetMyOrder = (data) => ({
  type: SET_ORDERS,
  data,
});

export const actionResetMyOrder = () => ({
  type: RESET_ORDERS,
});

export const actionCancelOrder = (id, callback) => ({
  type: CANCEL_ORDER,
  id,
  callback,
});
