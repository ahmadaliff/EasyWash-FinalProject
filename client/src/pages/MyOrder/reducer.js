import { produce } from 'immer';
import { RESET_ORDERS, SET_ORDERS } from './constants';

export const initialState = {
  orders: null,
};

export const myOrderReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case SET_ORDERS:
        draft.orders = action.data;
        break;
      case RESET_ORDERS:
        return initialState;
    }
  });

export default myOrderReducer;
