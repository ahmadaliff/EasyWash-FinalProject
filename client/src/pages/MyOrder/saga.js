import { setLoading, showPopup } from '@containers/App/actions';
import { apiCancelOrder, apiGetMyOrder } from '@domain/api';
import intlHelper from '@utils/intlHelper';
import toast from 'react-hot-toast';
import { call, put, takeLatest } from 'redux-saga/effects';
import { actionSetMyOrder } from './action';
import { CANCEL_ORDER, GET_ORDERS } from './constants';

function* sagaGetMyOrder() {
  yield put(setLoading(true));
  try {
    const response = yield call(apiGetMyOrder);
    yield put(actionSetMyOrder(response.data));
  } catch (error) {
    if (error?.response?.status === 400 || error?.response?.status === 404 || error?.response?.status === 401) {
      toast.error(error.response.data.message);

      toast.error(intlHelper({ message: error.response.data.message }));
    } else {
      yield put(showPopup());
    }
  }
  yield put(setLoading(false));
}

function* sagaCancelOrder({ id, callback }) {
  yield put(setLoading(true));
  try {
    const response = yield call(apiCancelOrder, id);
    toast.success(intlHelper({ message: response.message }));
    yield call(callback);
  } catch (error) {
    if (error?.response?.status === 400 || error?.response?.status === 404 || error?.response?.status === 401) {
      toast.error(error.response.data.message);

      toast.error(intlHelper({ message: error.response.data.message }));
    } else {
      yield put(showPopup());
    }
  }
  yield put(setLoading(false));
}

export default function* myOrderSaga() {
  yield takeLatest(GET_ORDERS, sagaGetMyOrder);
  yield takeLatest(CANCEL_ORDER, sagaCancelOrder);
}
