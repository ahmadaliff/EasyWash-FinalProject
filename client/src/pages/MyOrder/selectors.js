import { createSelector } from 'reselect';

import { initialState } from '@pages/MyOrder/reducer';

const selectMyOrderState = (state) => state.myOrder || initialState;

export const selectMyOrder = createSelector(selectMyOrderState, (state) => state.orders);
