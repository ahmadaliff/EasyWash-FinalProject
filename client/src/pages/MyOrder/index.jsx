import PropTypes from 'prop-types';
import { useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { FormattedMessage, FormattedNumber } from 'react-intl';

import { Button } from '@mui/material';
import { ArrowBack, DryCleaning } from '@mui/icons-material';

import { selectMyOrder } from '@pages/MyOrder/selectors';
import { actionGetMyOrder, actionResetMyOrder } from '@pages/MyOrder/action';

import classes from '@pages/MyOrder/style.module.scss';

const MyOrder = ({ orders }) => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  useEffect(() => {
    if (!orders) dispatch(actionGetMyOrder());
    return () => {
      if (orders) dispatch(actionResetMyOrder());
    };
  }, [dispatch, navigate, orders]);
  return (
    <main className={classes.mainWrap} data-testid="my-order">
      <div className={classes.header}>
        <h3>
          <FormattedMessage id="app_detail_order" />
        </h3>
        <Button
          variant="contained"
          className={classes.backButton}
          startIcon={<ArrowBack />}
          onClick={() => navigate(-1)}
          data-testid="back-button"
        >
          <FormattedMessage id="app_back" />
        </Button>
      </div>
      {orders?.map((order, key) => (
        <div className={classes.order} key={key}>
          <DryCleaning />
          <div className={classes.orderInfoWrap}>
            <hr />
            <div className={classes.orderInfo}>
              <div>
                <p>
                  <b>
                    <FormattedMessage id="app_order_id" /> : {order?.id}
                  </b>
                </p>
                <p>
                  <FormattedMessage id={order?.status} />
                </p>
              </div>
              <div>
                <p>
                  <FormattedNumber
                    value={order?.totalPrice}
                    // eslint-disable-next-line react/style-prop-object
                    style="currency"
                    currency="IDR"
                    minimumFractionDigits={0}
                  />
                </p>
                <div
                  className={classes.buttonDetail}
                  onClick={() => navigate(`/user/order/status/${order?.id}`)}
                  data-testid="button-order-detail"
                >
                  <FormattedMessage id="app_detail_order" />
                </div>
              </div>
            </div>
            <hr />
          </div>
        </div>
      ))}
    </main>
  );
};

MyOrder.propTypes = {
  orders: PropTypes.array,
};
const mapStateToProps = createStructuredSelector({
  orders: selectMyOrder,
});
export default connect(mapStateToProps)(MyOrder);
