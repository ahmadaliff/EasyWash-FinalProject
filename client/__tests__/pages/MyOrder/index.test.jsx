import MyOrder from '@pages/MyOrder';
import { render, fireEvent, act } from '@utils/testHelper';

let wrapper;
const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

jest.mock('@pages/MyOrder/selectors', () => ({
  selectMyOrder: () => [
    {
      status: 'app_pickUp',
      userId: 1,
      totalRows: 1,
      id: 1,
      totalPrice: 20000,
    },
  ],
}));

beforeEach(() => {
  act(() => {
    wrapper = render(<MyOrder />);
  });
});

describe('My Order Page', () => {
  test('render', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('my-order')).toBeInTheDocument();
  });
  test('navigate back when back button click', async () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('back-button');
    await act(async () => {
      fireEvent.click(button);
    });
    expect(mockNavigate).toHaveBeenCalledWith(-1);
  });
  test('shouldbe navigate to detail order', async () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-order-detail');
    act(() => {
      fireEvent.click(button);
    });
    expect(mockNavigate).toHaveBeenCalledWith('/user/order/status/1');
  });
  test('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
