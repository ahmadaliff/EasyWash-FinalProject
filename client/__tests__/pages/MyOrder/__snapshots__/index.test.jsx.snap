// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Order Page Matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="mainWrap"
        data-testid="my-order"
      >
        <div
          class="header"
        >
          <h3>
            Detail pesanan
          </h3>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium backButton css-sghohy-MuiButtonBase-root-MuiButton-root"
            data-testid="back-button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowBackIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                />
              </svg>
            </span>
            Kembali
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="order"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DryCleaningIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.56 11.36 13 8.44V7c0-.55-.45-1-1-1s-1-.45-1-1 .45-1 1-1 1 .45 1 1h2c0-1.84-1.66-3.3-3.56-2.95-1.18.22-2.15 1.17-2.38 2.35-.3 1.56.6 2.94 1.94 3.42v.63l-6.56 2.92c-.88.38-1.44 1.25-1.44 2.2v.01C3 14.92 4.08 16 5.42 16H7v6h10v-6h1.58c1.34 0 2.42-1.08 2.42-2.42v-.01c0-.95-.56-1.82-1.44-2.21zM18.58 14H17v-1H7v1H5.42c-.23 0-.42-.19-.42-.43 0-.17.1-.32.25-.38l6.75-3 6.75 3c.15.07.25.22.25.39 0 .23-.19.42-.42.42z"
            />
          </svg>
          <div
            class="orderInfoWrap"
          >
            <hr />
            <div
              class="orderInfo"
            >
              <div>
                <p>
                  <b>
                    Id Pesanan
                     : 
                    1
                  </b>
                </p>
                <p>
                  Diambil
                </p>
              </div>
              <div>
                <p>
                  Rp 20.000
                </p>
                <div
                  class="buttonDetail"
                  data-testid="button-order-detail"
                >
                  Detail pesanan
                </div>
              </div>
            </div>
            <hr />
          </div>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="mainWrap"
      data-testid="my-order"
    >
      <div
        class="header"
      >
        <h3>
          Detail pesanan
        </h3>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium backButton css-sghohy-MuiButtonBase-root-MuiButton-root"
          data-testid="back-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowBackIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              />
            </svg>
          </span>
          Kembali
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="order"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DryCleaningIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19.56 11.36 13 8.44V7c0-.55-.45-1-1-1s-1-.45-1-1 .45-1 1-1 1 .45 1 1h2c0-1.84-1.66-3.3-3.56-2.95-1.18.22-2.15 1.17-2.38 2.35-.3 1.56.6 2.94 1.94 3.42v.63l-6.56 2.92c-.88.38-1.44 1.25-1.44 2.2v.01C3 14.92 4.08 16 5.42 16H7v6h10v-6h1.58c1.34 0 2.42-1.08 2.42-2.42v-.01c0-.95-.56-1.82-1.44-2.21zM18.58 14H17v-1H7v1H5.42c-.23 0-.42-.19-.42-.43 0-.17.1-.32.25-.38l6.75-3 6.75 3c.15.07.25.22.25.39 0 .23-.19.42-.42.42z"
          />
        </svg>
        <div
          class="orderInfoWrap"
        >
          <hr />
          <div
            class="orderInfo"
          >
            <div>
              <p>
                <b>
                  Id Pesanan
                   : 
                  1
                </b>
              </p>
              <p>
                Diambil
              </p>
            </div>
            <div>
              <p>
                Rp 20.000
              </p>
              <div
                class="buttonDetail"
                data-testid="button-order-detail"
              >
                Detail pesanan
              </div>
            </div>
          </div>
          <hr />
        </div>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
